### ISSUE TITLE
[Feature Req.] Better Exception Handling

### ISSUE THREAD

**Issue by steven-murray** [21cmfast/21cmFAST#134]
Currently we have exception handling based on a *convention*: any C API-level function should return an int (the error status), and should have the entire body wrapped in a `Try{} Catch{}`. This is fine, and works. However, given that it's a pattern, we could shift it to be enforced via some kind of wrapping C or Python function. 

I'm not extremely well-versed in the limitations of C functions, but I feel like some kind of variadic function could be defined that takes a pointer to a function as input, and wraps the call in the Try/Catch, and adds the status etc. In Python, we'd just write a wrapper which calls the basic function _through_ the wrapper.

**Comment by BradGreig** [comment]
Hey @steven-murray, have you had any more thoughts on the implementation of this?

I have been thinking it might be worth my time sitting down and going through the code updating error messages etc. At the moment, an error message is output through the logging, then every parameter is dumped and then finally the uninformative error message (e.g. `21cmFAST does not support this combination of parameters.`).

I suspect in most instances people will not scroll up beyond all the dumped numbers to see the more informative error that is output through `LOG_ERROR`.

Is there a way for the `LOG_ERROR` message to be passed to python so that we get something more informative at the end of error message.

The other way I was thinking of this is to make the current exception coding considerably more granular to try and encompass a broader range of potential issues that might make the output python exception more informative.

If you think increased granularity in the exceptions is useful regardless, I can go ahead and do it in the meantime.

**Comment by steven-murray** [comment]
Hi @BradGreig. No, I have not had time to think about this more. 

It would be awesome to pass back the logged error message to python, but I think it will be very tricky. We can think about making the return type of the `Throw{}` statement to be a `char*`, in which case we can pass back arbitrary strings as the message. But I have a feeling this will be complicated.

I think more granularity is always good. When thinking about this, we should think about `Exception` class inheritance in Python. The reason I have it set up the way it is now is that `ParameterError` stuff should *not* cause a full exit when running MCMC -- you just give -inf for that likelihood and move on. Whereas fatal errors should cause the whole thing to crash. If you want to make the parameter errors more specific, they should inherit from `ParameterError`, so that we can still catch on `ParameterError` in 21cmMC.

**Comment by BradGreig** [comment]
Hi @steven-murray, ok I'll take a look at how this is currently implemented for `ParameterError` and keep it in mind if I choose to expand the error reporting. Might get to this in the next couple of days.

**Comment by BradGreig** [comment]
Hey @steven-murray, so I was looking into this today. I think in general I disagree with how the `ParameterError` stuff is used. I think in quite a few instances (say for example encountering infinities or NaN values) this shouldn't be assigned a -inf for a likelihood. The primary reason is that this could easily be caused not by the input parameters but rather numerical problems (and I'd say more likely the latter). In which case I think this should definitely cause a full exit so that these errors can be caught, reported and fixed (where possible).

For example, there could be a numerical error (just in how the code is implemented) that causes a small region of parameter space to return -inf instead of an actual sensible likelihood. The resultant posteriors would then look weird, and it wouldn't be obvious why. This in fact might have already been seen by @qyx268 whereby just some unfortunate edge case stuff with the `PhotonCons` stuff caused a series of models to throw this error and return -inf to the MCMC. Whereas otherwise, these should be reasonably normal parameter choices.

Thus, in my increasing of the granularity I am going to set most of the current `ParameterError` cases to in fact be `FatalCError`s instead. While I can concede that these might be a bit annoying initially, having these identified, reported and fixed should ultimately result in the code becoming a lot more robust to these issues

**Comment by qyx268** [comment]
@BradGreig is right. We should do this properly and crash the program when it fails. Right now, providing warnings is just a quick get-around, which is okay when using the emcee sampler but does cause some trouble with nested samplers.

Possible duplication to #189

**Comment by qyx268** [comment]
a second thought... perhaps it's worth to have something like `-f` option, so that the user can still ignore these errors and the program returns a -inf likelihood instead of crashing? I'm just thinking of the PhotonCon error which can be fixed ny extending z range for the analytic solution. Then the user followes the instuction and extends the range, but it crashes again (in rare cases) because the z range needs to be further extended... This can be annoying unless we can find a better solution for this types of "error".

**Comment by steven-murray** [comment]
I agree with @qyx268. In fact, I had thought I had already implemented something of the sort, but just checked and I haven't. The idea would be that the user gets to specifically choose if they want to ignore such "soft" errors and keep going. It'd be _really_ annoying if there's just one point that 21cmFAST can't quite handle on the outer edge of the parameter space and it blows everything up after 1000 iterations.
